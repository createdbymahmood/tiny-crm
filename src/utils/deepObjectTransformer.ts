import { isArray, isObject, map, transform } from 'lodash';

interface DeepObjectTransformerParams<T> {
    obj: T;
    predicate: (value: any) => boolean;
    transformer: (value: any) => any;
}

/**
 *
 * @param obj - object to transform
 * @param predicate - function that returns true if value should be transformed
 * @param transformer - function that transforms value
 * @returns object with transformed values
 */
export function deepObjectTransformer<T>({
    obj,
    predicate,
    transformer,
}: DeepObjectTransformerParams<T>): T {
    try {
        return transform(
            obj as unknown[],
            (result: any, value: any, key: string) => {
                if (predicate(value)) {
                    result[key] = transformer(value);
                } else if (isArray(value)) {
                    result[key] = map(value, item =>
                        deepObjectTransformer({
                            obj: item,
                            predicate,
                            transformer,
                        }),
                    );
                } else if (isObject(value)) {
                    result[key] = deepObjectTransformer({
                        obj: value,
                        predicate,
                        transformer,
                    });
                } else {
                    result[key] = value;
                }
            },
        );
    } catch (error) {
        return obj;
    }
}
