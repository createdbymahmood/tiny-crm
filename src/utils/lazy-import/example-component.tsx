export const ExampleComponent: React.FC = () => {
    return <div>Example Component to test</div>;
};
