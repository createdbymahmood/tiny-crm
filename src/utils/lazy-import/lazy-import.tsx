import * as React from 'react';

import {Pending} from '@/app/general';

import {retry} from '../retry';

interface LazyImportOptions {
  fallback: React.ReactNode;
}

type Unpromisify<T> = T extends Promise<infer P> ? P : never;

type ImportFunc<U> = () => Promise<{default: U}>;

export const lazyImport = <
  T extends Promise<any>,
  U extends React.ComponentType<any>,
>(
  importFunc: () => T,
  selectorFunc?: (s: Unpromisify<T>) => U,
  opts: LazyImportOptions = {fallback: null},
) => {
  // eslint-disable-next-line fp/no-let
  let lazyFactory: ImportFunc<U> = () => retry<{default: U}>(importFunc);

  if (selectorFunc) {
    lazyFactory = () =>
      importFunc().then(module => ({default: selectorFunc(module)}));
  }

  const LazyComponent = React.lazy(lazyFactory);

  return (props: React.ComponentProps<U>): JSX.Element => (
    <React.Suspense fallback={opts.fallback ?? <Pending />}>
      {/* <Page> */}
      <LazyComponent {...props} />
      {/* </Page> */}
    </React.Suspense>
  );
};
