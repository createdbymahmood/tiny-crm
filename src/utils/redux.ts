export function providesList<
    R extends { id?: string | number }[],
    T extends string,
>(resultsWithIds: R | undefined, tagType: T) {
    return resultsWithIds
        ? [
              { type: tagType, id: 'LIST' },
              ...resultsWithIds.map(({ id }) => ({ type: tagType, id })),
          ]
        : [{ type: tagType, id: 'LIST' }];
}
