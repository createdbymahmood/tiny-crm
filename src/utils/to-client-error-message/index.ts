export * from './to-client-error-message';
