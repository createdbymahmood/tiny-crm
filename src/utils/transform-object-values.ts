import {isArray, isObject, map, transform} from 'lodash';

interface TransformObjectValuesParams<T> {
    obj: T;
    predicate: (value: any) => boolean;
    transformer: (value: any) => any;
}

/**
 *
 * @param obj - object to transform
 * @param predicate - function that returns true if value should be transformed
 * @param transformer - function that transforms value
 * @returns object with transformed values
 */
export function transformObjectValues<T>({
    obj,
    predicate,
    transformer,
}: TransformObjectValuesParams<T>): T {
    try {
        return transform(
            obj as unknown[],
            (result: any, value: any, key: string) => {
                if (predicate(value)) {
                    result[key] = transformer(value);
                } else if (isArray(value)) {
                    result[key] = map(value, item =>
                        transformObjectValues({
                            obj: item,
                            predicate,
                            transformer,
                        }),
                    );
                } else if (isObject(value)) {
                    result[key] = transformObjectValues({
                        obj: value,
                        predicate,
                        transformer,
                    });
                } else {
                    result[key] = value;
                }
            },
        );
    } catch (error) {
        return obj;
    }
}
