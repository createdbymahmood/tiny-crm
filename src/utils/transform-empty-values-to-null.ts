import {isEmpty, isNull, isString} from 'lodash';

import {deepObjectTransformer} from '@/utils/deep-object-transformer';

/**
 *
 * @param obj  - object to transform
 * @returns object with empty values transformed to null
 */
export function transformEmptyValuesToNull<T>(obj: T) {
    return deepObjectTransformer<T>({
        obj,
        predicate: value =>
            isNull(value) || (isString(value) && isEmpty(value)),
        transformer: () => null,
    });
}
