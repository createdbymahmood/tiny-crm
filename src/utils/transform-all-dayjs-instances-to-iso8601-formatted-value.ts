import {ISO_8601_DATE_FORMAT} from '@configs/constants';
import type {Dayjs} from 'dayjs';
import dayJS from 'dayjs';

import {transformObjectValues} from '@/utils/transform-object-values';

/**
 *
 * @param obj - object to transform
 * @returns object with all dayjs instances transformed to ISO 8601 formatted value
 */

export function transformAllDayjsInstancesToIso8601FormattedValue<T>(obj: T) {
  return transformObjectValues<T>({
    obj,
    predicate: value => dayJS.isDayjs(value),
    transformer: (value: Dayjs) => value.format(ISO_8601_DATE_FORMAT),
  });
}
