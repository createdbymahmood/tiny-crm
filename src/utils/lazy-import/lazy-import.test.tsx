import {render, screen} from '@testing-library/react';
import {describe, expect, it} from 'vitest';

import {sleep} from '@/utils/sleep';

import {lazyImport} from './lazy-import';

const loadingIndicator = 'Loading...';

describe('lazyImport', () => {
    it('should render without any props', () => {
        const LazyComponent = lazyImport(() => import('./example-component'));
        const {container} = render(<LazyComponent />);

        expect(container).toMatchSnapshot();
    });

    it('displays fallback element while loading', async () => {
        const LazyComponent = lazyImport(
            () => sleep(1000).then(() => import('./example-component')),
            undefined,
            {fallback: <div>{loadingIndicator}</div>},
        );
        render(<LazyComponent />);
        const loadingElement = await screen.findByText(loadingIndicator);

        expect(loadingElement).not.toBeUndefined();
        expect(loadingElement).not.toBeNull();
    });
});
