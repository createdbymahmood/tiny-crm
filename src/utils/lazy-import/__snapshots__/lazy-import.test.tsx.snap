// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lazyImport > should render without any props 1`] = `
<div>
  <div
    aria-busy="true"
    aria-live="polite"
    class="ant-spin ant-spin-spinning css-dev-only-do-not-override-k7429z"
  >
    <span
      class="ant-spin-dot ant-spin-dot-spin"
    >
      <i
        class="ant-spin-dot-item"
      />
      <i
        class="ant-spin-dot-item"
      />
      <i
        class="ant-spin-dot-item"
      />
      <i
        class="ant-spin-dot-item"
      />
    </span>
  </div>
</div>
`;
